src/app/qgisapp.cpp

```diff
@@ -3240,16 +3240,9 @@ bool QgisApp::addVectorLayers( const QStringList &theLayerQStringList, const QSt
         //set friendly name for datasources with only one layer
         QStringList sublayers = layer->dataProvider()->subLayers();
         QStringList elements = sublayers.at( 0 ).split( ':' );
-        if ( layer->storageType() != "GeoJSON" )
-        {
-          while ( elements.size() > 4 )
-          {
-            elements[1] += ':' + elements[2];
-            elements.removeAt( 2 );
-          }
 
-          layer->setLayerName( elements.at( 1 ) );
-        }
+        layer->setLayerName( QString( "%1 %2 %3" ).arg( layer->name() ).arg( elements.at( 1 ) ).arg( elements.at( 3 ) ) );
```

Comment:
To be on the safe side an assertion could be introduced to get a meaningful feedback on dev machines in case the assumption does not always hold true.

    Q_ASSERT( elements.size() >= 4 )